src/tools/start_stop.ml

```diff
@@ -43,24 +46,22 @@ class virtual base ~(on_start : unit -> unit) ~(on_stop : unit -> unit) =
 
     method transition_to (s : state) =
       match (s, state) with
-        | `Started, `Stopped ->
+        | `Started, `Stopped | `Started, `Idle ->
             self#start;
             on_start ();
             state <- `Started
+        | `Started, `Started -> ()
         | `Stopped, `Started ->
-            if self#stype = Source.Infallible then
-              raise
-                Lang_values.(
-                  Runtime_error
-                    {
-                      kind = "input";
-                      msg = Some "Input is infallible and cannot be stopped";
-                      pos = [];
-                    });
             self#stop;
             on_stop ();
             state <- `Stopped
-        | _ -> ()
+        | `Stopped, `Idle -> state <- `Stopped
+        | `Stopped, `Stopped -> ()
+        | `Idle, `Started ->
+            self#stop;
+            on_stop ();
+            state <- `Idle
+        | `Idle, `Stopped | `Idle, `Idle -> ()
```

Comment:
Shouldn't we have an error (or at least a message in the logs) for `Idle -> Stopped`?

Reply:
No, `Idle` is an automated stop that was not ordered by the user. Asking to `Stop` from there just makes the `Stop` explicit. The difference is that the output will not restart automatically once the underlying source gets ready again.